dbms/src/Storages/MergeTree/ReplicatedMergeTreeCleanupThread.cpp

```diff
@@ -108,11 +109,117 @@ void ReplicatedMergeTreeCleanupThread::clearOldLogs()
     LOG_DEBUG(log, "Removed " << entries.size() << " old log entries: " << entries.front() << " - " << entries.back());
 }
 
+namespace
+{
+
+struct RequiredStat
+{
+    Int64 ctime;
+    int numChildren;
+
+    RequiredStat() = default;
+    RequiredStat(const RequiredStat &) = default;
+    explicit RequiredStat(const zkutil::Stat & s) : ctime(s.ctime), numChildren(s.numChildren) {};
+    explicit RequiredStat(Int64 ctime_) : ctime(ctime_) {}
+};
+
+}
+
+
+class ReplicatedMergeTreeCleanupThread::NodesStatCache : public std::map<String, RequiredStat>
+{
+};
+
+
+struct ReplicatedMergeTreeCleanupThread::NodeWithStat
+{
+    String node;
+    RequiredStat stat;
+
+    NodeWithStat() = default;
+    NodeWithStat(const String & node_, const RequiredStat & stat_) : node(node_), stat(stat_) {}
+
+    static bool greaterByTime (const NodeWithStat & lhs, const NodeWithStat & rhs)
+    {
+        return (lhs.stat.ctime != rhs.stat.ctime) ? lhs.stat.ctime > rhs.stat.ctime : lhs.node > rhs.node;
+    }
+};
+
 
 void ReplicatedMergeTreeCleanupThread::clearOldBlocks()
 {
     auto zookeeper = storage.getZooKeeper();
 
+    std::vector<NodeWithStat> timed_blocks;
+    getBlocksSortedByTime(zookeeper, timed_blocks);
+
+    if (timed_blocks.empty())
+        return;
+
+    /// Use ZooKeeper's first node (last according to time) timestamp as "current" time.
+    Int64 current_time = timed_blocks.front().stat.ctime;
+    Int64 time_threshold = std::max(0L, current_time - static_cast<Int64>(storage.data.settings.replicated_deduplication_window_seconds));
```

Comment:
Bug, because ZooKeeper have time in milliseconds, but `replicated_deduplication_window_seconds` is in seconds.